dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/OpenAPIVersionConfigImpl.java

```diff
@@ -0,0 +1,282 @@
+            Tr.warning(tc,
+                       MessageConstants.OPENAPI_VERSION_INVALID_CWWK06181W,
+                       versionString,
+                       getSupportedVersions(),
+                       this.configuredVersion);
+            return;
+        }
+
+        if (!isSupported(version.get())) {
+            this.configuredVersion = defaultVersion();
+            Tr.warning(tc,
+                       MessageConstants.OPENAPI_VERSION_NOT_SUPPORTED_CWWK06182W,
+                       version.get(),
+                       getSupportedVersions(),
+                       this.configuredVersion);
+            return;
+        }
+
+        // Configured version is acceptable
+        this.configuredVersion = version.get();
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            Tr.event(this, tc, "OpenAPI version set", this.configuredVersion);
+        }
+    }
+
+    @Override
+    public void applyConfig(OpenAPI model) {
+        if (!ProductInfo.getBetaEdition()) {
+            // Never apply config if we're not the beta edition
+            return;
+        }
+
+        // Get the version out of the model
+        Optional<OpenAPIVersion> modelVersion = OpenAPIVersion.parse(model.getOpenapi());
+
+        // Check if it needs replaced
+        boolean replace = modelVersion.map(this::requiresReplacement)
+                                      .orElse(true);
+
+        if (replace) {
+            // Update the version in the model
+            model.setOpenapi(getReplacementVersion());
+        }
+    }
+
+    /**
+     * Returns the default value for the OpenAPI version
+     *
+     * @return the default OpenAPI version
+     */
+    protected OpenAPIVersion defaultVersion() {
+        return VERSION_30;
+    }
+
+    /**
+     * Returns whether the given OpenAPI version is supported by this version of the mpOpenAPI feature
+     *
+     * @param version the version to check
+     * @return {@code true} if this is a supported version, {@code false} otherwise
+     */
+    protected boolean isSupported(OpenAPIVersion version) {
+        return version.major == 3 && version.minor == 0;
+    }
+
+    /**
+     * Lists the supported OpenAPI spec versions for output in error messages.
+     *
+     * @return a comma separated list of OpenAPI spec versions which are supported
+     */
+    protected String getSupportedVersions() {
+        return "3.0, 3.0.x";
+    }
+
+    /**
+     * Returns whether an OpenAPIVersion parsed from the model should be replaced, based on the configured version
+     *
+     * @param modelVersion the version parsed from the model
+     * @return {@code true} if the version should be updated, {@code false} otherwise
+     */
+    protected boolean requiresReplacement(OpenAPIVersion modelVersion) {
+        if (modelVersion.major == configuredVersion.major
+            && modelVersion.minor == configuredVersion.minor
+            && modelVersion.patch >= 0
```

Comment:
This if statement will always fall into the else if modelVersion.patch is -1 , even if both modelVersion and configueredVersion are completely identical. I want to confirm this is correct.

Reply:
This is correct. The modelVersion [must be three digits](https://spec.openapis.org/oas/v3.1.0.html#versions) whereas the configured version can be two.

I'll break this condition up and add comments.